ipapython/ipautil.py

```diff
@@ -282,6 +282,37 @@ def write_tmp_file(txt):
 
     return fd
 
+
+@contextmanager
+def concurrent_open(path, mode='r'):
+    """Ensure that complete file is read/written
+
+    Works only for r, rb, w, w+, wb, wb+ modes. Default is r.
+
+    In read mode behaves the same as build-in open. In write mode all data are
+    written to the temporary file first and then moved to target path.
+    This approach ensures that reading will be performed on complete file and
+    writting will result in complete or none file written at all.
+    """
+    if mode in ('w', 'w+', 'wb', 'wb+'):
+        with tempfile.NamedTemporaryFile(
+                mode=mode, prefix=path, delete=False) as temp_file:
```

Comment:
Regarding the prefix you're right, I haven't thought about relative names. I know that rename works only inside single FS otherwise raises error.

Regarding yielding `NamedTemporaryFile` instance from concurrent_open I'm not sure I understand you. Do you want to say that `NamedTemporaryFile.__exit__` is called when `yield temp_file` is executed on line 300? I don't agree with that because the context haven't finished yet. The problem I see not with line 300 is that I should enclose it with `try: finally:`.